src/features/courses/course-controller.ts

```diff
@@ -2522,6 +2634,243 @@ class CourseController {
+        return topic.topicTypeId === 2;
+    }
+
+    /*
+    * Determine if a user is allowed to view a question 
+    * If the question belongs to an assessment, a studentTopicAssessmentInfoId is REQUIRED
+    */
+    async userCanViewQuestionId(user: User, questionId: number, studentTopicAssessmentInfoId?: number): Promise<boolean> {
+        if (user.roleId === Role.PROFESSOR || user.roleId === Role.ADMIN) return true;
+        const question = await courseRepository.getQuestion({ id: questionId });
+        const topic = await question.getTopic();
+        if (topic.topicTypeId === 1) {
+            return (topic.startDate.toMoment().isBefore(moment()));
+        } else if (topic.topicTypeId === 2) {
+            if (_.isNil(studentTopicAssessmentInfoId)) throw new WrappedError('Cannot tell if a user can view this question when a version id is not supplied');
+            const topicInfo = await topic.getTopicAssessmentInfo();
+            const studentTopicInfo = await this.getStudentTopicAssessmentInfoById(studentTopicAssessmentInfoId);
+            const topicIsLive = moment().isBetween(studentTopicInfo.startTime.toMoment(), studentTopicInfo.endTime.toMoment());
+            if (topicIsLive) {
+                return true;
+            } else {
+                // strike that -- reverse it ;)
+                return !topicInfo.hideProblemsAfterFinish;
+            }
+        }
+        // we should never get here - but if we do, then the answer is NO
+        return false;
+    };
+
+    scoreAssessment = (results: SubmittedAssessmentResultContext[]): ScoreAssessmentResult => {
+        let totalScore = 0;
+        let bestVersionScore = 0;
+        let bestOverallVersion = 0;
+        const problemScores: Array<number> = [];
+
+        results.forEach((result: SubmittedAssessmentResultContext) => {
+            const { questionResponse, grade, instance, weight } = result;
+            totalScore += weight * questionResponse.problem_result.score;
+            bestVersionScore += weight * instance.scoreForBestVersion;
+            bestOverallVersion += weight * grade.bestScore;
+            problemScores[result.instance.problemNumber] = result.questionResponse.problem_result.score * result.weight;
+        });
+
+        problemScores[0] = totalScore; // scores indexed by problem number, total at [0]
+        return {problemScores, bestVersionScore: bestVersionScore, bestOverallVersion};
+    }
+
+    async submitAssessmentAnswers(studentTopicAssessmentInfoId: number, wasAutoSubmitted: boolean): Promise<SubmitAssessmentAnswerResult> {
+        const studentTopicAssessmentInfo = await this.getStudentTopicAssessmentInfoById(studentTopicAssessmentInfoId);
+        if (studentTopicAssessmentInfo.numAttempts >= studentTopicAssessmentInfo.maxAttempts) {
+            throw new WrappedError('Cannot submit assessment answers when there are no attempts remaining'); // sanity check, shouldn't happen
```

Comment:
How should this be returned to the user, the way it is now will be an internal server error, however it seems like it should be a bad request